src/plugins/data/common/index_patterns/index_patterns/index_pattern.ts

```diff
@@ -104,8 +103,7 @@ export class IndexPattern implements IIndexPattern {
     this.title = spec.title || '';
     this.timeFieldName = spec.timeFieldName;
     this.sourceFilters = spec.sourceFilters;
-    this.attributes = spec.attributes;
-    this.fields.replaceAll(Object.values(this.getFieldSpecs(spec.fields)));
+    this.fields.replaceAll(Object.values(spec.fields || {}));
     this.type = spec.type;
     this.typeMeta = spec.typeMeta;
```

Comment:
@mattkime There's a problem with this approach, it is never using the values that are stored in the index pattern's saved object attributes property, they are just stored there. So the `customName` values stored in the `fields` map are displayed. However, if you refresh the fields list, those values get lost. So if we do not want to use a method like `getFieldSpecs`, we would need to extend `this.fields.replaceAll`?

Reply:
They're added to the field spec - https://github.com/elastic/kibana/pull/82400/commits/bbd65b05fa8905f72137cff156c8483e702c1ed5#diff-444560d042fab9c4f51faa1db6cb9f2576ec920c449cebb0af6300f7b86683afR306

I thought I verified the functionality but I'll double check.

Reply:
@mattkime you're right! did only a quick check wha I saw is this:

![image](https://user-images.githubusercontent.com/463851/98040685-f6015000-1e20-11eb-861a-b4f85b9ec63b.gif)

But when I've reloaded the page, my custom name was displayed, so persistence works (guess it's just a minor issue to fix showing the customNames when refreshing fields, but it's a bit too late for me to have a closer look now)

Reply:
I'll take a look

Reply:
Pushed fixes. Of note, its necessary to manually serialize and deserialize `attributes` otherwise empty props won't overwrite existing props. This is a problem when removing a customName.

I'll need to look at this with fresh eyes tomorrow.

Reply:
@mattkime I did some research and I wonder if there have been changes recently in the index pattern API? 
I can confirm that if you're not serializing and deserializing the attributes, removed props are no persisted, and so they are not removed. It's because how the ES Update API works, it's doing partial updated in this case? Lens team had the same problem and fixed it with a workaround:

https://github.com/elastic/kibana/blob/fe3b0538ffdeaea1b8468d2532f368e297756f38/x-pack/plugins/lens/public/persistence/saved_object_store.ts#L72

Wonder if it's worth to go into that direction or just use strings